src/pgen/cluster/agn_feedback.cpp

```diff
@@ -0,0 +1,370 @@
+                        magnetic_fraction_ >= 0,
+                    "AGN feedback energy fractions must be non-negative.");
+
+  /////////////////////////////////////////////////////
+  // Read in or calculate jet velocity and temperature. Either and or both can
+  // be defined but they must satify
+  //
+  // v_jet = sqrt( 2*(eps*c^2 - (1-eps)*e_jet) )
+  //
+  // With real, non-negative values for v_jet and e_jet
+  /////////////////////////////////////////////////////
+
+  kinetic_jet_velocity_ = NAN;
+  kinetic_jet_temperature_ = NAN;
+
+  Real kinetic_jet_e;
+
+  const auto units = hydro_pkg->Param<Units>("units");
+
+  if (pin->DoesParameterExist("problem/cluster/agn_feedback", "kinetic_jet_velocity")) {
+    kinetic_jet_velocity_ =
+        pin->GetReal("problem/cluster/agn_feedback", "kinetic_jet_velocity");
+  }
+  if (pin->DoesParameterExist("problem/cluster/agn_feedback",
+                              "kinetic_jet_temperature")) {
+    kinetic_jet_temperature_ =
+        pin->GetReal("problem/cluster/agn_feedback", "kinetic_jet_temperature");
+
+    const Real He_mass_fraction = pin->GetReal("hydro", "He_mass_fraction");
+    const Real H_mass_fraction = 1.0 - He_mass_fraction;
+    const Real mu = 1 / (He_mass_fraction * 3. / 4. + (1 - He_mass_fraction) * 2);
+    const Real gam = pin->GetReal("hydro", "gamma");
+    const Real gm1 = (gam - 1.0);
+    kinetic_jet_e = units.k_boltzmann() * kinetic_jet_temperature_ /
+                    (mu * units.atomic_mass_unit() * gm1);
+  }
+
+  if (std::isnan(kinetic_jet_velocity_) && std::isnan(kinetic_jet_temperature_)) {
+    // Both velocity and temperature are missing, assume 0K temperature
+    kinetic_jet_velocity_ = units.speed_of_light() * sqrt(2 * (efficiency_));
+    kinetic_jet_temperature_ = 0;
+    kinetic_jet_e = 0;
+    std::cout << "### WARNING Kinetic jet velocity nor temperature not specified. "
+                 "Assuming 0K temperature jet"
+              << std::endl;
+  } else if (std::isnan(kinetic_jet_velocity_)) {
+    // Velocity is missing, compute it from e_jet
+    kinetic_jet_velocity_ = sqrt(2 * (efficiency_ * SQR(units.speed_of_light()) -
+                                      (1.0 - efficiency_) * kinetic_jet_e));
+  } else if (std::isnan(kinetic_jet_temperature_)) {
+    // Temperature is missing, compute e_jet and T_jet from v_jet
+    kinetic_jet_e =
+        (efficiency_ * SQR(units.speed_of_light()) - 0.5 * SQR(kinetic_jet_velocity_)) /
+        (1 - efficiency_);
+    const Real He_mass_fraction = pin->GetReal("hydro", "He_mass_fraction");
+    const Real H_mass_fraction = 1.0 - He_mass_fraction;
+    const Real mu = 1 / (He_mass_fraction * 3. / 4. + (1 - He_mass_fraction) * 2);
+    const Real gam = pin->GetReal("hydro", "gamma");
+    const Real gm1 = (gam - 1.0);
+    kinetic_jet_temperature_ =
+        (kinetic_jet_e * mu * units.atomic_mass_unit() * gm1) / units.k_boltzmann();
```

Comment:
Reminder to update computing the temperature in this new code that was written about the same time as the mu-update.